File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Call/Views/LocalVideoView/LocalVideoView.m
//
//  LocalVideoView.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "LocalVideoView.h"

@interface LocalVideoView()
//MARK: - Properties
@property (weak, nonatomic) AVCaptureVideoPreviewLayer *videoLayer;
@property (strong, nonatomic) UIView *containerView;

@end

@implementation LocalVideoView
//MARK: - Life Cycle
- (void)dealloc {

    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (instancetype)initWithPreviewlayer:(AVCaptureVideoPreviewLayer *)layer {
    self = [super initWithFrame:CGRectZero];
    if (self) {
        
        self.videoLayer = layer;
        layer.videoGravity = AVLayerVideoGravityResizeAspectFill;
        
        self.containerView = [[UIView alloc] initWithFrame:self.bounds];
        self.containerView.backgroundColor = [UIColor clearColor];
        [self.containerView.layer insertSublayer:self.videoLayer atIndex:0];
        
        [self insertSubview:self.containerView atIndex:0];
    }
    
    return self;
}

- (void)willMoveToSuperview:(UIView *)newSuperview {
    [super willMoveToSuperview:newSuperview];
    [self updateOrientationIfNeeded];
}

- (void)layoutSubviews {
    [super layoutSubviews];
    
    self.containerView.frame = self.bounds;
    self.videoLayer.frame = self.bounds;
}

//MARK: - Private Methods
- (void)updateConstraints {
    [super updateConstraints];
    self.containerView.frame = self.bounds;
    self.videoLayer.frame = self.bounds;
    
}

- (void)updateOrientationIfNeeded {
    AVCaptureConnection *previewLayerConnection = self.videoLayer.connection;
    UIInterfaceOrientation interfaceOrientation = UIApplication.sharedApplication.windows.firstObject.windowScene.interfaceOrientation;
    AVCaptureVideoOrientation videoOrientation = (AVCaptureVideoOrientation)interfaceOrientation;
    
    BOOL isVideoOrientationSupported = [previewLayerConnection isVideoOrientationSupported];
    if (isVideoOrientationSupported
        && previewLayerConnection.videoOrientation != videoOrientation) {
        [previewLayerConnection setVideoOrientation:videoOrientation];
    }
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Calls/Views/LocalVideoView/LocalVideoView.swift
//
//  LocalVideoView.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 04.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit
import AVKit

class LocalVideoView: UIView {
    //MARK: - Properties
    var videoLayer: AVCaptureVideoPreviewLayer?
    
    lazy private var containerView: UIView = {
        let containerView = UIView(frame: bounds)
        containerView.backgroundColor = UIColor.clear
        insertSubview(containerView, at: 0)
        return containerView
    }()
    
    //MARK: - Life Circle
    public init(previewlayer layer: AVCaptureVideoPreviewLayer) {
        super.init(frame: CGRect.zero)
        videoLayer = layer
        videoLayer?.videoGravity = .resizeAspectFill
        containerView.layer.insertSublayer(layer, at:0)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()

        containerView.frame = bounds
        videoLayer?.frame = bounds
    }
    
    override func willMove(toSuperview newSuperview: UIView?) {
        super.willMove(toSuperview: newSuperview)
        updateOrientationIfNeeded()
    }
    
    //MARK: - Internal Methods
    private func updateOrientationIfNeeded() {
        let previewLayerConnection = videoLayer?.connection
        let isVideoOrientationSupported = previewLayerConnection?.isVideoOrientationSupported
        
        guard let interfaceOrientation = UIApplication.shared.windows.first?.windowScene?.interfaceOrientation,
              let videoOrientation = AVCaptureVideoOrientation(rawValue: interfaceOrientation.rawValue),
            isVideoOrientationSupported == true,
            previewLayerConnection?.videoOrientation != videoOrientation else {
                return
        }
        previewLayerConnection?.videoOrientation = videoOrientation
    }
}
